# Mapping and Assembly

## In this exercise, we will de novo assemble a viral genome from a publically available dataset and will use read mapping to validate our assembly.  We will:

* Download a dataset from the SRA
* Convert the dataset from SRA -> FASTQ format
* Find and download a genome sequence from NCBI
* Create a bowtie index from the genome sequence (necessary to map reads to it)


## Download an SRA dataset

We will download one of the NGS datasets reported in [this paper](http://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1004900)

This dataset was generated by performing shotgun sequencing of total RNA from the liver of a boa constrictor that was diagnosed with [inclusion body disease](https://en.wikipedia.org/wiki/Inclusion_body_disease). The dataset is composed of reads from host RNAs and from viral RNAs.  

First, let's setup a directory (folder) in which to work.  Lines that start with a `#` symbol are comment lines and are just for your information.  Open the terminal app on your laptop, then type these commands:

```
# change (move) to your home directory, if not already there
cd

# make a new directory
mkdir boa_sra

# move to that directory
cd boa_sra

# double check you are in the directory you think you are:
pwd
```

To get the dataset, open a browser and enter this address, the pubmed page for the dataset's paper:

https://www.ncbi.nlm.nih.gov/pubmed/25993603

Scroll down and find the 'Related information' section of the bottom right of the page.  Click on the SRA link.  This shows the 146 datasets associated with this paper.  Search for `snake_7`.  Note that the reads in this dataset are already trimmed.  Note the run # (SRR #) for this dataset: SRR1984309

To download the dataset to your computer, you can run this command from your terminal window:

```
curl -O ftp.ncbi.nlm.nih.gov:/sra/sra-instant/reads/ByRun/sra/SRR/SRR198/SRR1984309/SRR1984309.sra
```

*5 sec over wifi*

See [this page](http://www.ncbi.nlm.nih.gov/books/NBK158899/) for more information about how this URL was determined.  Here is a [tool](https://github.com/stenglein-lab/stenglein_lab_scripts/blob/master/download_sra_data) for downloading SRA data that will automate this process.

Confirm that you downloaded the file:

```
# you should see a file named SRR1984039.sra that is 22 Mb
ls -lh
```


## Convert SRA->FASTQ

SRA datasets arrive in SRA format.  We need them to be in a more usable format, like [FASTQ](https://en.wikipedia.org/wiki/FASTQ_format).   

We will use the [SRA toolkit](https://www.ncbi.nlm.nih.gov/books/NBK158900/) to convert from SRA to FASTQ.

```
# fastq-dump converts SRA->FASTQ
fastq-dump --split-files SRR1984309.sra

# rename files using the mv (move) command
mv SRR1984309_1.fastq SRR1984309_R1.fastq
mv SRR1984309_2.fastq SRR1984309_R2.fastq
```

*~1 sec*

## Download the boa constrictor (mtDNA) genome.

The dataset we downloaded is from boa constrictor liver RNA.  We are going to map the reads in the dataset to the boa constrictor mtDNA genome sequence to demonstrate read mapping.  We'll use the [bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml) mapper.

#### First we need to actually get the genome sequence to which we will map:

* Goto NCBI Taxonomy database: https://www.ncbi.nlm.nih.gov/taxonomy
* Search for "boa constrictor"
* Click on boa constrictor link
* Click on genome (1) link
*
* Click on 'See also 1 organelle- and plasmid-only records matching your search/
* Click on 'NC_007398.1' RefSeq link

#### As is usually the case in bioinformatics, there is more than one way to do what we want:

* Option 1: Download from website using browser
 * download sequence from NCBI website
 * Send->complete record->file->format[fasta]
 * Move this FASTA file to your ~/boa_sra directory using the Finder or the command line

* Option 2: download the sequence in Geneious.  
 * Copy the accession # from your browswer page.  
 * Open Geneious.  Goto the NCBI->Nucleotide section.  
 * Search for the accession (NC_007398.1) .  
 * Create a new folder in Geneious.  
 * Drag the boa mtDNA sequence to this new folder.
 * Note the nice annotation.
 * Export the sequence in FASTA format.  File->Export->Selected Documents->Fasta sequences/alignment format.  Click through options.  

#### Create a bowtie index

Now we will create a bowtie2 index from our boa mtDNA sequence.  Indexing pre-processes a sequence to make it faster to map to later.

In your terminal window run these commands:

```
# confirm that the fasta file is there
ls -lh    # should see: NC_007398.fasta

# run bowtie2-build to make index
# this command takes 2 arguments: 
# (1) the name of the fasta file containing the sequence(s) you will index
# (2) the name of the index (can be whatever you want)

bowtie2-build NC_007398.fasta NC_007398

# confirm that you built the index.  You should see a bunch of files named ending in bt2, like NC_007398.3.bt2
ls -lh
```




